### PR TITLE
feat: Update Dockerfile to add support to arbitrary user ids

### PR DESCRIPTION
**Pull request by edsoncsouza**
About this change - What it does

This change updates the Dockerfile to handle dynamic User ID (UID) allocation in compliance with OpenShift best practices. It modifies file and directory permissions to ensure compatibility with OpenShift's security requirements, allowing containers to run with arbitrary UIDs while maintaining the required access to essential directories.

References: #983 

Why this way

To solve the issue of dynamic UID handling, the proposed approach involves modifying permissions so that the container user can run with a dynamically assigned UID, which OpenShift uses for security purposes. Specifically, we change directory permissions and add the 'karapace' user to group '0', ensuring necessary access without granting root privileges. This adheres to OpenShift's recommendations for secure, multi-tenant container environments, allowing seamless and secure container operation.

### CODE REVIEW

File: container/Dockerfile [hunk 1]
```diff
@@ -11,9 +11,10 @@ ENV PIP_REQUIRE_VIRTUALENV=true
 # Install golang needed by extensions
 ENV GO_VERSION=1.21.0
 ENV PATH="/usr/local/go/bin:${PATH}"
-RUN wget --progress=dot:giga "https://go.dev/dl/go${GO_VERSION}.linux-$(dpkg --print-architecture).tar.gz" \
- && tar -C /usr/local -xzf "go${GO_VERSION}.linux-$(dpkg --print-architecture).tar.gz" \
- && rm "go${GO_VERSION}.linux-$(dpkg --print-architecture).tar.gz"
+RUN wget --progress=dot:giga \
```
Comment (jjaakola-aiven) [inline container/Dockerfile:14]: nit: unnecessary change.

File: container/Dockerfile [hunk 2]
```diff
@@ -29,35 +30,34 @@ COPY . /build/karapace-repo
 WORKDIR /build/karapace-repo
 RUN --mount=type=cache,target=/root/.cache/pip \
     if [ -z "${KARAPACE_VERSION}" ]; then \
-        PRETEND_VERSION="$(python -c 'from src.karapace import version; print(version.__version__)')"; \
+    PRETEND_VERSION="$(python -c 'from src.karapace import version; print(version.__version__)')"; \
```
Comment (jjaakola-aiven) [inline container/Dockerfile:33]: nit: unnecessary change.

File: container/Dockerfile [hunk 3]
```diff
@@ -29,35 +30,34 @@ COPY . /build/karapace-repo
...
     else \
-        PRETEND_VERSION=$KARAPACE_VERSION; \
+    PRETEND_VERSION=$KARAPACE_VERSION; \
```
Comment (jjaakola-aiven) [inline container/Dockerfile:35]: nit: unnecessary change.

File: container/Dockerfile [hunk 4]
```diff
@@ -29,35 +30,34 @@ COPY . /build/karapace-repo
...
+RUN useradd --system --gid 0 karapace \
+    && mkdir /opt/karapace /opt/karapace/runtime /var/log/karapace \
+    && chgrp -R 0  /opt/karapace /opt/karapace/runtime /var/log/karapace \
+    && chmod -R g+rwX /opt/karapace
 
 # Install protobuf compiler.
 ARG PROTOBUF_COMPILER_VERSION="3.12.4-1+deb11u1"
 RUN apt-get update \
- && apt-get install --assume-yes --no-install-recommends \
+    && apt-get install --assume-yes --no-install-recommends \
```
Comment (jjaakola-aiven) [inline container/Dockerfile:51]: nit: unnecessary change.

File: container/Dockerfile [hunk 5]
```diff
@@ -29,35 +30,34 @@ COPY . /build/karapace-repo
...
     protobuf-compiler=$PROTOBUF_COMPILER_VERSION \
- && rm -rf /var/lib/apt/lists/*
+    && rm -rf /var/lib/apt/lists/*
```
Comment (jjaakola-aiven) [inline container/Dockerfile:53]: nit: unnecessary change.

**Review by jjaakola-aiven — COMMENTED** [review]
I'd prefer the formatting related changes in separate commit and having the actual change in one.

**Review by eliax1996 — APPROVED** [review]
LGTM except the nits.
Can we add a comment with the doc reference in the docker file?
I think [reading this](https://docs.redhat.com/en/documentation/openshift_container_platform/4.17/html/images/creating-images#use-uid_create-images) can be useful to understand why.

**Comment by edsoncsouza** [comment]
I've made the corrections
Thanks for the reviews

**Review by jjaakola-aiven — APPROVED** [review]
